Automattic/jetpack
# Staging: deprecate staging logic and replace with safe mode and dev site

This PR changes the way we handle staging and development sites. The idea is  to allow sync on all staging sites, deprecate the usage of Jetpack staging constant, and disable the ability to migrate the connection when site enters Safe mode on a staging or dev site. Please read about details here: pf5801-Pb-p2

## Proposed changes:

* In the Status class we are deprecating the previously used is_staging_site method that was used for staging sites (using a staging constant or a known staging domain) and for sites in the safe mode .
* Instead, there is a new method for sites using WP_ENVIRONMENT_TYPE staging or development that gets used in js package for Safe mode prompt. Development sites don't get to migrate their connection to the new URL and they can only stay in Safe mode or create a new connection.
* Ther is also a new method that blocks sync when sites are in Safe mode (IDC)

### Other information:


There are new unit tests for the new method and I updated the old tests.

## Jetpack product discussion



pf5801-Pb-p2

## Does this pull request change what data or activity we track or use?



No

## Testing instructions:






* You will need a connected JN site. Make sure you install the Jetpack Debug tools plugin too. Get its blog id.
* Use debug tools to trigger IDC.
* Load My Jetpack. A Safe mode should pop up looking as usual.
* Set the site to staging mode by adding define( 'WP_ENVIRONMENT_TYPE', 'staging'); in wp-config or by using a filter. Something like this in Code snippets should do:
```
add_filter( 'jetpack_is_development_site',  '__return_true' );
```
* Reload My Jetpack. The Safe mode should look different (no migration option).
* Try using Stay in Safe Mode. The orange safe mode button should appear in the admin bar. Click on it to get the Safe mode back.
* Break REST API. Using Disable REST API plugin is good for this. Find the `/jetpack/v4/identity-crisis/confirm-safe-mode` on the list and disable it for admin. Don't forget to save :-) 
* Try picking the Stay in Safe mode again. You should see the `Could not stay in safe mode`  error.
* Reeanable API and pick the Create fresh connection option. The site should reconnect to its old blog id.
* The site should still be in development mode so test it. Sync should work and no Staging mode notices should show up in Jetpack dashboard. Try if stats are updated, or full sync, do some stuff and make sure they show up in the activity log on WPcom.
